### ISSUE TITLE
Question: why we need to add EarlGrey.framework to DYLD_INSERT_LIBRARIES

### ISSUE THREAD

**Issue by superb** [google/EarlGrey#653]
I've read https://github.com/google/EarlGrey/blob/master/docs/ifaq.md and it explained why. 

I am wondering if this is necessary, so I did some experiments:

1. I added a module initializer to EarlGray.m:

__attribute__((constructor))
static void initializer(int argc, char** argv, char** envp) {
  static int initialized = 0;
  if (!initialized) {
    // Initialization code.
    initialized = 1;
  }
}

and setup a breakpoint in it.

2. Run any test in ExampleTest, when the breakpoint is hit, it shows the following call stack:

* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
  * frame #0: 0x0000000151817b2f (*hide test name*)`initializer(argc=5, argv=0x00007ffee21268a0, envp=0x00007ffee21268d0) at EarlGrey.m:34
    frame #1: 0x00000001336d7c16 dyld_sim`ImageLoaderMachO::doModInitFunctions(ImageLoader::LinkContext const&) + 420
    frame #2: 0x00000001336d7e46 dyld_sim`ImageLoaderMachO::doInitialization(ImageLoader::LinkContext const&) + 40
    frame #3: 0x00000001336d36da dyld_sim`ImageLoader::recursiveInitialization(ImageLoader::LinkContext const&, unsigned int, char const*, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 330
    frame #4: 0x00000001336d2898 dyld_sim`ImageLoader::processInitializers(ImageLoader::LinkContext const&, unsigned int, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 134
    frame #5: 0x00000001336d292c dyld_sim`ImageLoader::runInitializers(ImageLoader::LinkContext const&, ImageLoader::InitializerTimingList&) + 74
    frame #6: 0x00000001336c9956 dyld_sim`dyld::runInitializers(ImageLoader*) + 82
    frame #7: 0x00000001336cf076 dyld_sim`dlopen + 803
    frame #8: 0x000000013e24aa47 libdyld.dylib`dlopen + 86
    frame #9: 0x0000000136b00808 CoreFoundation`_CFBundleDlfcnLoadBundle + 152
    frame #10: 0x0000000136b00694 CoreFoundation`_CFBundleLoadExecutableAndReturnError + 324
    frame #11: 0x00000001342b4ec1 Foundation`-[NSBundle loadAndReturnError:] + 520
    frame #12: 0x00000001337a3860 IDEBundleInjection`__XCBundleInjection + 798
    frame #13: 0x00000001336d7c16 dyld_sim`ImageLoaderMachO::doModInitFunctions(ImageLoader::LinkContext const&) + 420
    frame #14: 0x00000001336d7e46 dyld_sim`ImageLoaderMachO::doInitialization(ImageLoader::LinkContext const&) + 40
    frame #15: 0x00000001336d36da dyld_sim`ImageLoader::recursiveInitialization(ImageLoader::LinkContext const&, unsigned int, char const*, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 330
    frame #16: 0x00000001336d2898 dyld_sim`ImageLoader::processInitializers(ImageLoader::LinkContext const&, unsigned int, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 134
    frame #17: 0x00000001336d292c dyld_sim`ImageLoader::runInitializers(ImageLoader::LinkContext const&, ImageLoader::InitializerTimingList&) + 74
    frame #18: 0x00000001336c715d dyld_sim`dyld::initializeMainExecutable() + 166
    frame #19: 0x00000001336cac98 dyld_sim`dyld::_main(macho_header const*, unsigned long, int, char const**, char const**, char const**, unsigned long*) + 4412
    frame #20: 0x00000001336c63d4 dyld_sim`start_sim + 136
    frame #21: 0x000000013c40cded dyld`dyld::useSimulatorDyld(int, macho_header const*, char const*, int, char const**, char const**, char const**, unsigned long*, unsigned long*) + 2200
    frame #22: 0x000000013c40a7a3 dyld`dyld::_main(macho_header const*, unsigned long, int, char const**, char const**, char const**, unsigned long*) + 436
    frame #23: 0x000000013c4063d4 dyld`dyldbootstrap::start(macho_header const*, int, char const**, long, macho_header const*, unsigned long*) + 453
    frame #24: 0x000000013c4061d2 dyld`_dyld_start + 54

3. Also add a break point at the beginning main function of App under test, and confirm:
  a. it is hit later than loading EarlGray.framework
  b. check EarlGray.framework did get loaded by dyld

So looks like once test target is linked against EarlGrey.framework, it get loaded at a very early stage when dyld is loading libraries for AUT. Is library loaded through DYLD_INSERT_LIBRARIES get loaded earlier than this? If true, is there anything EarlGrey is tracking even before main function get running?

**Comment by tirodkar** [comment]
Though @khandpur can explain this much better, the order of loading is not guaranteed to us and can vary depending on the iOS / Xcode Version. All we care about is that EarlGrey is definitely loaded before the app under test is, so we track all its state changes. To ensure this, we use DYLD_INSERT_LIBRARIES.

**Comment by superb** [comment]
Thanks for the quick response!

I am asking because I am using BUCK to setup the Xcode project/workspace and it doesn't provide a way to specify command line parameters or environment variable. So I just tried without setting up DYLD_INSERT_LIBRARIES and surprisingly it still works.

Also I am testing on Xcode 9 + iOS 11, not sure if this still works on older version Xcode and iOS

**Comment by tirodkar** [comment]
That's good to know that it works fine on iOS 11. We had previously seen this issue with iOS 9 / Xcode 7 or so. I'm not sure if BUCK allows changing the Schemes, which is also where DYLD_INSERT_LIBRARIES can be changed from.

**Comment by superb** [comment]
Yeah, maybe it's worth mentioning in the manual setup doc that hacking library loading with DYLD_INSERT_LIBRARIES may not be necessary (at least on iOS 11), so people don't spend extra time setting it up (and spend much more time if their Xcode project is populated from some config files)

Just want to confirm it is OK to skip that step and thanks for the quick reply! Feel free to close if no further action is needed :)

**Comment by tirodkar** [comment]
That would be a good idea. Would you like to add it for us?

**Comment by superb** [comment]
Sure! Haven't contributed to projects on github yet but I will figure it out later!

**Comment by tirodkar** [comment]
@superb, quick question - are you running the tests on device. On a mac, the dynamic linker remembers the path to the framework being linked, which is probably giving you a false positive.

**Comment by superb** [comment]
No, I run my test on simulator. Will do a test on device later and give more details!
